include/deflation.h

```diff
@@ -0,0 +1,244 @@
+#pragma once
+
+#include <invert_quda.h>
+#include <vector>
+#include <complex_quda.h>
+
+namespace quda {
+
+  /**
+     This struct contains all the metadata required to define the
+     deflated solvers.  
+   */
+  struct DeflationParam {
+
+    /** This points to the parameter struct that is passed into QUDA.
+	We use this to set (per-level) parameters */
+    QudaEigParam  &eig_global;
+
+    /** Buffer for Ritz vectors 
+        For staggered: we need to reduce dimensionality of each component?
+     */
+    ColorSpinorField *RV;  
+    
+    /** Inverse Ritz values*/
+    double *invRitzVals;
+
+     /** The Dirac operator to use for spinor deflation operation */
+    DiracMatrix &matDeflation;
+
+    /** Host  projection matrix (e.g. eigCG VH A V) */
+    Complex *matProj; 
+
+    /** projection matrix leading dimension */
+    int ld;                 
+
+    /** projection matrix full (maximum) dimension (nev*deflation_grid) */
+    int tot_dim; 
+
+    /** current dimension (must match rhs_idx: if(rhs_idx < deflation_grid) curr_nevs <= nev * rhs_idx) */           
+    int cur_dim;
+
+    /** use inverse Ritz values for deflation (for the best performance) */            
+    bool use_inv_ritz;
+
+    /** Where to compute Ritz vectors */
+    QudaFieldLocation location;
+
+    /** Filename for where to load/store the deflation space */
+    char filename[100];
+
+    DeflationParam(QudaEigParam &param, ColorSpinorField *RV,  DiracMatrix &matDeflation, int cur_dim = 0) : eig_global(param), RV(RV), matDeflation(matDeflation), 
+             cur_dim(cur_dim), use_inv_ritz(false), location(param.location) {
+
+        if(param.nk == 0 || param.np == 0 || (param.np % param.nk != 0)) errorQuda("\nIncorrect deflation space parameters...\n");
+        //redesign: param.nk => param.nev, param.np => param.deflation_grid*param.nev;
+        tot_dim      = param.np;
+        ld           = ((tot_dim+15) / 16) * tot_dim;
+        //allocate deflation resources:
+        matProj      = new Complex[ld*tot_dim];
+        invRitzVals  = new double[tot_dim];
+
+        //Check that RV is a composite field:
+        if(RV->IsComposite() == false) errorQuda("\nRitz vectors must be contained in a composite field.\n");
+
+        cudaHostRegister(matProj,ld*tot_dim*sizeof(Complex),cudaHostRegisterDefault);
+     }
+
+     ~DeflationParam(){
+        cudaHostUnregister(matProj);
+        if(matProj) delete[]  matProj;
+        if(invRitzVals)       delete[]  invRitzVals;
+     }
+#if 0 //we cannot reshape the set here.
```

Comment:
If this code is for the future, can you add a comment to that effect?  Or should it just be deleted?